### DISCUSSION TITLE
choose a language

### DISCUSSION THREAD

**Discussion by IRoN-cP** [NuSkooler/enigma-bbs#478] [Ideas]
Hello @all !
My name is IRoN and I've been the sysop of the Crazy Paradise BBS since the late 80's. To date, my system ran / runs on an Amiga. However, there are now more and more reasons for me to set up a new system. The election goes to Enigma. It is basically the same as my Amiga system, but the programming language is completely different. In my system it is a batch language in a kind of basic. So now I have to completely relearn everything. I tried to create a selection menu to give the user the possibility to choose a language. In my case German, English.
NuSkooler emailed me to open a ticket here. I hope I'm in the right place and someone can help me.

Hopeful greetings from Berlin
IRON

**Comment by NuSkooler** [comment]
Hi IRoN!

Not sure if you saw it yet, but I responded in-BBS as well with a few details.

We can use this as a discussion area of localization, then create ticket(s).

**First off the current state of things**
* No current support for language/localization selection. That's why were here!
* Currently **most** strings are found in art files (e.g. `mymenu.ans`)
* Some strings are found within the source. Minimal, but they are there.
* Additional strings in `menu.hjson` and `theme.hjson` (for example, menu `items[]`)

**Adding Localization**
So some things we'd need to tackle:
1. Storing the user's localization selection should be easy: Store in a user property.
2. ACS can already be used to check user properties
3. We need a way to localize the hard-coded strings found in the source
4. We need a way to localize the `menu.hjson` and `theme.hjson` entries. This one may be slightly tricky. A lot of localization systems (ie: GNU gettext) take the input "my string" and look it up in a table for a particular language. In ENiGMA however, "my string" may have color codes embedded such as "|04my |05string".

Thoughts?

**Reply by IRoN-cP** [reply]
Hi NuSkooler,maybe we don't need to meditate so "extremely" in the various files. 
I think a selection of the theme should be enough.
**English
German
.**....
There you can create everything in the desired language. The few small texts that are in there somewhere else are not that important at first, since a lot of things in a German mailbox are in English anyway.
So a small script to select the language should be enough. Perhaps there is still the possibility to change this in the user settings.

would that be easy to do?

Greetings IRoN

**Comment by NuSkooler** [comment]
IRoN --

If that's acceptable for the first iteration, I think you are right, and 
wouldn't require any changes actually:

1) Copy e.g. a English theme

2) Update all the art/etc. to be German

3) In the new theme's 'theme.hjson' file, override any 'items', 'text', 
etc. to be German

With that, users could select "Theme - English" vs "Theme - German" in 
their config. Creating a mod that allows "language selection" but in 
fact really is just a theme selection is quite easy.

Let me know if that works for you -- I'll still keep the localization 
ticket open and ponder it more so we can get something better in the 
future.

Guten tag!

Bryan / NuSkooler

On 2023-07-27 15:12, IRoN-cP wrote:

> Hi NuSkooler,maybe we don't need to meditate so "extremely" in the 
> various files.
> I think a selection of the theme should be enough.
> English
> German
> .....
> There you can create everything in the desired language. The few small 
> texts that are in there somewhere else are not that important at first, 
> since a lot of things in a German mailbox are in English anyway.
> So a small script to select the language should be enough. Perhaps 
> there is still the possibility to change this in the user settings.
> 
> would that be easy to do?
> 
> Greetings IRoN
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe 
> [2].
> You are receiving this because you commented.Message ID: 
> ***@***.***>


Links:
------
[1] 
https://github.com/NuSkooler/enigma-bbs/discussions/478#discussioncomment-6566603
[2]

**Reply by IRoN-cP** [reply]
Yes, I and the user can definitely handle that. 
I copied the luciano directory as cp_ger. I'm currently making all the changes there. 
In config.hjson I changed the two defaults to cp_ger. If I have everything ready, I will create a cp_eng based on the cp_ger. 
Currently I'm still learning how the MCI codes etc work. I've been working with the Amiga BBS system for 30 years and I'm kind of starting from scratch. but it's fun and i see the progress.
If I change anything else, please let me know.
Ich wünsche die ebenfalls einen guten Tag :)

**Comment by NuSkooler** [comment]
IRoN --

Great! ENiG does quite a few things differently than even "oldschool" PC 
based systems, so there is certainly a learning curve either way. Let me 
know how it goes and if you want to build a simple "language selection" 
mod and need help I can do that as well.

Bryan

On 2023-07-27 16:13, IRoN-cP wrote:

> Yes, I and the user can definitely handle that.
> I copied the luciano directory as cp_ger. I'm currently making all the 
> changes there.
> In config.hjson I changed the two defaults to cp_ger. If I have 
> everything ready, I will create a cp_eng based on the cp_ger.
> Currently I'm still learning how the MCI codes etc work. I've been 
> working with the Amiga BBS system for 30 years and I'm kind of starting 
> from scratch. but it's fun and i see the progress.
> If I change anything else, please let me know.
> Ich wünsche die ebenfalls einen guten Tag :)
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe 
> [2].
> You are receiving this because you commented.Message ID: 
> ***@***.***>


Links:
------
[1] 
https://github.com/NuSkooler/enigma-bbs/discussions/478#discussioncomment-6567328
[2]
